active/0000-addon-instrumentation-experimental-hooks.md

```diff
@@ -0,0 +1,400 @@
+write addons that do things like summarize and report build performance
+information.
+
+- see https://github.com/ember-cli/ember-cli/issues/6349 for additional context.
+- see https://github.com/ember-cli/ember-cli/pull/6606 for an experimental
+  implementation.
+
+# Motivation
+
+Build performance is important to users.  We want to enable users to:
+
+1. Easily discover which portions of their build are costly;
+2. Be able to summarize and report build information in an addon;
+3. Be able to write addons that analyze build performance instrumentation so
+   that they can more easily help diagnose build performance issues in projects
+   to which they do not have direct access.  This is of particular interest to
+   @ember-cli/core &c.
+
+In order to provide these hooks to enable iteration and experimentation prior to
+making firm commitments to format, this rfc propose to initially expose them as
+experiments (see the experiments section below).
+
+# Detailed design
+
+## Experiments
+
+Experiments live in `lib/experiments/index.js`.  Unlike feature flags, there is
+no need to strip them from production.  Experiments allow us to provide power
+user features that are not fully stable without their resorting to private API
+usage.
+
+Experiments are available only in canary builds.  This is achieved by only
+including `lib/experiements/index.js` in canary, and making it the entry point
+for all experiments.
+
+## Instrumentation Hook
+
+We have already a build instrumentation hook as an
+experiment in https://github.com/ember-cli/ember-cli/pull/6546
+
+A more encompassing instrumentation hook is implemented in
+https://github.com/ember-cli/ember-cli/pull/6606
+
+
+The goal of this RFC is:
+
+  1. To make the concept of experiments supported and explicit
+  2. To promote this particular experiment to public API
+
+### Enabling Instrumentation
+
+Instrumentation is enabled if either the environment variable `BROCCOLI_VIZ` is
+set to `1` or if `EMBER_CLI_INSTRUMENTATION` is set to `1`.
+
+If `BROCCOLI_VIZ=1` then in addition to instrumentation hooks being invoked, a
+serialized form of the instrumentation information is written to disk, that is
+appropriate for consumption by [broccoli-viz](https://github.com/ember-cli/broccoli-viz) which is the current behaviour.
+
+### Instrumentation
+
+#### Hook
+
+An addon that implements `instrumentation` will have this hook invoked when
+instrumentation is enabled.
+
+```js
+module.exports = {
+  name: 'my-great-addon',
+
+  instrumentation(name, payload) {
+    // format of instrumentation payload outlined below
+  }
+};
+```
+
+##### name
+
+The `name` argument indicates what phase the instrumentation payload describes.
+In beta and released versions this will always be a string.
+
+On canary it could be a symbol from `lib/experiments` if we add more phases (eg
+more fine-grained phases) for instrumentation information.
+
+The initial set of phases this RFC advocates are:
```

Comment:
This is a great suggestion.  Quick summary for now

1. init is everything before actually running the command.  This means mostly `require`ing time, addon discovery &c.
2. command is the entire command, whatever the command is.  An important exception for `ember serve` is that we'll only include perf information about the last build, to avoid leaks.
3. build is the builder portion of a build command, or of a build within a serve command.
4. shutdown is everything from command to process exit.